### PR TITLE
fix(recover): require sustained gateway serving after recovery (#4710)

### PR DESCRIPTION
**Pull request by ericksoa**
## Summary

`nemoclaw <name> recover` declared success on a single gateway health probe. A wedged gateway (#4710) serves for ~20 seconds after logging ready and then drops its HTTP listener while the process stays alive — so recover reported a healthy gateway that was already on its way back to the wedge. This PR makes recovery require *sustained* serving and surfaces the wedge signature when it recurs.

## Related Issue

Related: #4710 (host-side hardening; helps existing sandboxes without an image rebuild — complements the sandbox-side fix).

## Changes

- `waitForRecoveredSandboxGateway` (`src/lib/actions/sandbox/process-recovery.ts`): after the first successful probe, wait a settle window (`NEMOCLAW_GATEWAY_RECOVERY_SETTLE_SECONDS`, default 25s, `0` disables) and require a confirm probe to still succeed before declaring recovery. Prints a one-line progress note so the wait doesn't read as a hang.
- New `collectGatewayWedgeDiagnostics`: on confirm failure, greps `/tmp/gateway.log` for the wedge signature (`config change requires gateway restart` / `gateway startup failed` / `Process will stay alive`) and prints the matching lines so the operator sees why the gateway is unreachable despite a live PID.
- Both functions take injectable probe/sleep/exec seams; tests cover settle-confirm success, the serve-then-drop wedge, the `0`-disable path, initial polling interplay, and diagnostics extraction/edge cases.

## Architecture

- **Health means proven serving, not existence.** Recovery already moved from pgrep to HTTP probes (#2342); the settle-confirm extends that same principle from point-in-time proof to sustained proof, matched to the wedge's ~20s time constant. Success criteria stay aligned with what the operator actually needs to be true.
- **Monolith reduction.** The wedge diagnostics live in a focused `gateway-wedge-diagnostics` module rather than growing the high-churn `process-recovery.ts` lifecycle monolith (the deterministic growth gate's direction). The sandbox exec is passed explicitly, keeping the import graph acyclic.
- **Trust boundaries.** Matched gateway-log lines are sandbox-writable bytes: control characters are neutralized and credential shapes redacted before anything reaches an operator terminal.
- **Workaround contract.** The module header documents the invalid OpenClaw state, the upstream source boundary, and the removal condition — this PR is detection/confirmation only and is deliberately scoped as the host-side complement to the sandbox-side fix (#5181); it does not claim the original #4710 HEALTHCHECK/marker acceptance.
- **Seams and parity.** Probe/sleep/exec are injectable in the established `execImpl` style, and tests run against the compiled `dist/` like the rest of the CLI suite, with the CLI-level settle test isolated in a focused file under the size budgets.

## Type of Change

- [x] Code change (feature, bug fix, or refactor)

## Verification

- [x] `npx prek run --all-files` passes — except the `test-cli` hook's two pre-existing, machine-bound failures that reproduce identically on a clean `main` checkout on this host (live e2e scenario `ubuntu-repo-cloud-openclaw`, timing-sensitive auto-pair keepalive retry); all other hooks pass.
- [x] `npm test` passes for the touched suites (`src/lib/actions/sandbox/process-recovery.test.ts`, `test/process-recovery.test.ts`).
- [x] Tests added or updated for new or changed behavior
- [x] No secrets, API keys, or credentials committed
- [ ] Docs updated for user-facing behavior changes — env var follows the existing undocumented `NEMOCLAW_GATEWAY_RECOVERY_WAIT_SECONDS`/`_POLL_INTERVAL_SECONDS` convention in the same function; happy to add a reference entry if preferred.

---
Signed-off-by: Aaron Erickson <aerickson@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Sandbox recovery now prints sanitized gateway "wedge" diagnostics on failure and supports an optional post-recovery "settle" confirmation (configurable; can be disabled).

* **Tests**
  * Expanded coverage for settle-confirm behavior, polling/timeouts, intermittent listener-drop scenarios, diagnostic extraction/sanitization, and a CLI test path that disables the settle delay for faster runs.

### CODE REVIEW

**Comment by ericksoa** [comment]
Re: PR Review Advisor findings —

**Scope (#4710 acceptance):** this PR is explicitly the *host-side complement*, not the closure of the original HEALTHCHECK/marker acceptance. The marker/HEALTHCHECK and sandbox-side prevention clauses are owned by #5116 (marker tied to the launch site) and #5181 (gateway.reload=hot pin + serving watchdog + HEALTHCHECK pattern). This PR only hardens `recover`/`connect --probe-only` so they cannot declare success inside the pre-wedge window, and it does not close #4710 on its own (the body links it as "Related", not "Fixes").

**Monolith growth:** addressed in c9c71cf4c — the wedge-diagnostics helpers moved out of `process-recovery.ts` into a focused `gateway-wedge-diagnostics.ts` module, bringing the net `process-recovery.ts` delta down to the settle-confirm itself.

**Source-of-truth / removal contract:** documented in the new module header — invalid state (OpenClaw gateway parks alive-but-deaf after a failed in-process restart), source boundary (OpenClaw run loop; upstream report in progress), and removal condition (an OpenClaw release whose failed restart exits non-zero so PID-wait supervisors respawn it).

**Log-line trust:** also in c9c71cf4c — matched gateway.log lines are sanitized before printing (terminal control characters stripped; bearer headers, key/token/secret/password assignments, and nvapi- keys redacted), with unit coverage.

**Runtime validation:** dispatched the advisor-required Vitest scenario `ubuntu-repo-docker-post-reboot-recovery` against this branch; the auto-dispatched `sandbox-operations-e2e` and `sandbox-survival-e2e` runs already passed.

File: test/cli/helpers.ts [hunk 1]
```diff
@@ -171,6 +171,11 @@ export function runWithEnv(
       HOME: fs.mkdtempSync(path.join(os.tmpdir(), "nemoclaw-cli-test-")),
       NEMOCLAW_HEALTH_POLL_COUNT: "1",
       NEMOCLAW_HEALTH_POLL_INTERVAL: "0",
+      // #4710: the post-recovery settle-confirm waits 25s by default; CLI
+      // tests disable it to stay fast. Settle behavior has dedicated
+      // coverage in process-recovery.test.ts and a targeted CLI test in
+      // connect-recovery.test.ts that overrides this with a short window.
+      NEMOCLAW_GATEWAY_RECOVERY_SETTLE_SECONDS: "0",
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply test/cli/helpers.ts:197]: Fixed — the comment now points at connect-recovery-settle.test.ts.

**Review by cv — APPROVED** [review]
